### PR TITLE
Conf tree

### LINKED ISSUE #3253: build documentation into `cylc.flow.cfgspec` modules
**Describe exactly what you would like to see in an upcoming release**

Take the configuration documentation out of cylc-doc and write it into the source code so that the documentation can be auto-generated.

* Ensures all configurations are documented and that no old configurations are accidentally left documented
* Prevents docs getting out-of-sync with cylc-flow
* Easier to work on configuration documentation in-place with cylc-flow commits

See "Auto-Documentation" section of https://github.com/cylc/cylc-doc/issues/11

*Tentatively added to the Cylc8 milestone to facilitate suite interface changes. Feel free to bump it back.*

**Additional context**

The global spec must be loaded before pretty much every Cylc command so it is important not to make the `SPEC`s too heavy.

Having said that would a little extra text cause problems?

* If not do it!
* If so investigate
  * Putting documentation in comments?
  * Putting documentation in another file?

**Question/Decision**

How should we encorporate documentation?

1. In the `SPEC` e.g.

```python
SPEC = {
    'meta': {
        'description': [
            VDR.V_STRING,
            '',
            '''
                A multi-line description of the suite. It can be retrieved
                at run time with the cylc show command.

                :type: multi-line string
                :default: (none)
            '''
        ],
```

2. In comments (yuck, and good luck parsing that) e.g.

```python
SPEC = {
    'meta': {
        'description': [VDR.V_STRING, '']
        #: A multi-line description of the suite. It can be retrieved
        #: at run time with the cylc show command.
        #:
        #: :type: multi-line string
        #: :default: (none)
```

3. In a separate documentation file?

```python
DOCSPEC = {
    'meta': (
        '''
        Section containing metadata items for this suite.
        Several items (title, description, URL) are pre-defined and
        are used by the GUI. Others can be user-defined and passed to
        suite event handlers to be interpreted according to your needs.
        For example, the value of a “suite-priority” item could determine
        how an event handler responds to failure events.
        ''',
        {
            'description': '''
                A multi-line description of the suite. It can be retrieved
                at run time with the cylc show command.

                :type: multi-line string
                :default: (none)
            ''',
```

4. Something alltogether different e.g:

```ini
[meta]
    [[description]]
        type = VDR.V_STRING
        default = ''
        description = """
            A multi-line description of the suite. It can be retrieved at
            run time with the cylc show command.

            type: multi-line string
            default: (none)
```

**Pull requests welcome!**

### PR DESCRIPTION
**Pull request by oliver-sanders**
These changes close #3253 

**Sibling PRs:**

* https://github.com/cylc/cylc-sphinx-extensions/pull/22
* https://github.com/cylc/cylc-doc/pull/121

**Highlights:**

* Define the "specification" for the suite and global configurations using "context nodes".
* Copy documentation in from cylc-doc (documentation will require review at a later date).

**Other Stuff:**

* Port some functional tests to pytest.
* Port some unittest to pytest.
* Remove some unnecessary escape chars from the old RST docs.

**Todo:**

* [x] Add metavars for `__MANY__` items.
* [x] Open a cylc-doc PR to autodocument this.

**Future work:**

* Consider changing the `options` list to a dictionary.
* Refactor the deprecation mechanism around the new configuration nodes.
* Review all documentation and configurations!!!
* Build examples into the parsec validator types.
  * Saves us having a `PT1H` example in every config which takes a duration as an arugment.
* Use SPEC items for retrieving config values (e.g. `self.config(SPEC['scheduling']['initial cycle point']`)
  * Allows us to detect when we are still relying on deprecated configs in Cylc.
  * Allows us to add hyperlinks to the reference material in error messages.
* Update all references configurations in the docs to use the `:cylc:conf'[a][b]c'` syntax.
  * Allows integrity checking of the documentation.
  * Gives you hyperlinks to the relevant sections.
  * Allows referencing of Cylc configurations from other Sphinx project (e.g. Rose).

**Context Node Fun:**

```python
>>> from cylc.flow.cfgspec.suite import SPEC

>>> print(SPEC.tree())
    [meta]
        description
        group
        title
        URL
        __MANY__
    [cylc]
        UTC mode
        cycle point format
        cycle point num expanded year digits
...

>>> for conf in SPEC['scheduling']: print(repr(conf))
[scheduling]initial cycle point
[scheduling]final cycle point
[scheduling]initial cycle point constraints
...


```

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] (master branch) I have opened a documentation PR at https://github.com/cylc/cylc-doc/pull/121

### CODE REVIEW

**Comment by oliver-sanders** [comment]
An early example of what the built documentation will look like (using the default Sphinx theme)

Untitled2

**Comment by wxtim** [comment]
> An early example of what the built documentation will look like (using the default Sphinx theme)

Is this example based on the code in this branch? If so, how?

File: cylc/flow/cfgspec/suite.py [hunk 1]
```diff
@@ -34,246 +33,1089 @@
...
+            This config item specifies a hard limit as a cycle interval
+            between the slowest and fastest tasks. See also
+            :cylc:conf:`[..]max active cycle points` which defines the limit
+            as a number of cyles.
+
+            Example: ``PT12H`` - for a 12 hour limit under ISO 8601 cycling.
+        ''')
+        Conf('max active cycle points', VDR.V_INTEGER, 3, desc='''
+            Runahead limiting prevents the fastest tasks in a suite from
+            getting too far ahead of the slowest ones, as documented in
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply cylc/flow/cfgspec/suite.py:265]: Configs gets alphabetically sorted before rendering the output in Sphinx so these two won't appear next to each other in the final docs.

File: cylc/flow/cfgspec/suite.py [hunk 2]
```diff
@@ -34,246 +33,1089 @@
...
+            the suite. Names may contain letters, digits, underscores, and
+            hyphens.  A namespace represents a group or family of tasks if
+            other namespaces inherit from it, or a task if no others inherit
+            from it.
+
+            Names may not contain colons (which would preclude use of
+            directory paths involving the registration name in ``$PATH``
+            variables). They may not contain the "." character (it will be
+            interpreted as the namespace hierarchy delimiter, separating
+            groups and names -huh?).
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply cylc/flow/cfgspec/suite.py:497]: Probably not! Unless it was special emphasis.

File: cylc/flow/parsec/config.py [hunk 3]
```diff
@@ -69,19 +74,30 @@ def expand(self):
...
-                            defs[key] = OrderedDictWithDefaults()
-                        stack.append((defs[key], spec[key]))
+                # for key, val in spec.items():
+                for node in spec:
+                    # if isinstance(val, dict):
+                    if not node.is_leaf():
+                        # if key not in defs:
+                        if node.name not in defs:
+                            defs[node.name] = OrderedDictWithDefaults()
+                        # stack.append((defs[key], spec[key]))
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply cylc/flow/parsec/config.py:84]: I left these in for dev purposes to help track down bugs, should have removed them.

**Comment by oliver-sanders** [comment]
Thanks for putting up the typos as suggestions, made things much easier :)

**Pull request merged**